src/config.rs

```diff
@@ -0,0 +1,1191 @@
+    /// rules that are checked when a connection attempts to send a message
+    Send(SendOperation),
+}
+type OptionalOperation = Option<Operation>;
+impl TryFrom<RuleAttributes> for OptionalOperation {
+    type Error = Error;
+
+    fn try_from(value: RuleAttributes) -> std::result::Result<Self, Self::Error> {
+        let has_connect = value.group.is_some() || value.user.is_some();
+        let has_own = value.own.is_some() || value.own_prefix.is_some();
+        let has_send = value.send_broadcast.is_some()
+            || value.send_destination.is_some()
+            || value.send_destination_prefix.is_some()
+            || value.send_error.is_some()
+            || value.send_interface.is_some()
+            || value.send_member.is_some()
+            || value.send_path.is_some()
+            || value.send_requested_reply.is_some()
+            || value.send_type.is_some();
+        let has_receive = value.receive_error.is_some()
+            || value.receive_interface.is_some()
+            || value.receive_member.is_some()
+            || value.receive_path.is_some()
+            || value.receive_sender.is_some()
+            || value.receive_requested_reply.is_some()
+            || value.receive_type.is_some();
+
+        let operations_count: i8 = vec![has_connect, has_own, has_receive, has_send]
+            .into_iter()
+            .map(i8::from)
+            .sum();
+
+        if operations_count > 1 {
+            return Err(Error::msg(format!("do not mix rule attributes for connect, own, receive, and/or send attributes in the same rule: {value:?}")));
+        }
+
+        if value.group.is_some() && value.user.is_some() {
+            return Err(Error::msg(format!(
+                "`group` cannot be combined with `user` in the same rule: {value:?}"
+            )));
+        }
+        if value.own.is_some() && value.own_prefix.is_some() {
+            return Err(Error::msg(format!(
+                "`own_prefix` cannot be combined with `own` in the same rule: {value:?}"
+            )));
+        }
+        if value.send_destination.is_some() && value.send_destination_prefix.is_some() {
+            return Err(Error::msg(format!("`send_destination_prefix` cannot be combined with `send_destination` in the same rule: {value:?}")));
+        }
+
+        if has_connect {
+            Ok(Some(Operation::Connect(ConnectOperation::from(value))))
+        } else if has_own {
+            Ok(Some(Operation::Own(OwnOperation::from(value))))
+        } else if has_receive {
+            Ok(Some(Operation::Receive(ReceiveOperation::from(value))))
+        } else if has_send {
+            Ok(Some(Operation::Send(SendOperation::from(value))))
+        } else {
+            Err(Error::msg(format!("rule must specify supported attributes for connect, own, receive, or send operations: {value:?}")))
+        }
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, PartialEq)]
+pub struct OwnOperation {
+    pub own: Option<Name>,
```

Comment:
I did want to bikeshed this naming patternn, actually :)

Operation is...  close but not quite right, because really these are parts of a policy that has to **match** an operation, these are not really the operation itself 

Not hard to rename later, but it annoys me, haha

Reply:
> Operation is... close but not quite right, because really these are parts of a policy that has to **match** an operation, these are not really the operation itself

That's true for everything in config but the context and module hierarchy (`config`) keeps things clear, IMO. Config will never have any operations so this can't be the actual operation.

Having said that, it's more like "NameOwnership" than "NameOwningOperation".

Reply:
This seems unresolved still.